### ISSUE TITLE
Timeout when request body is consumed before proxying

### ISSUE THREAD

**Issue by maxsnts** [microsoft/reverse-proxy#1412]
### Describe the bug
The simple fact of accessing the Request.Form in a Middleware, while Posting a form, will freeze the Core App.

### To Reproduce
Add this line to a middleware 
```c#
if (context.Request.HasFormContentType && context.Request.Form != null)
{ 
     //...
}
```

This will let GETs go through fine, but will lock the callint core application when posting a form. 
Tested while making a postback into a WebFormsApplication

Thank you

**Comment by Tratcher** [comment]
`Request.Form` has to actually read the request body which is an async operation. That itself shouldn't hang, but too many requests doing that can cause threadpool starvation.

Two recommendations:
- `context.Request.Form != null` isn't a meaningful check, the property always returns a non-null object (or throws).
- Call ReadFormAsync() the first time you access the form to avoid threadpool starvation.

(this is unrelated to YARP, it's a general aspnetcore question).

**Comment by maxsnts** [comment]
I'm sorry to press.
This is replicable on test applications with one request only. No meed for many requests.
We can change the check and try to count the items in the form collection. It will hang just the same.

**Comment by Tratcher** [comment]
Can you share the test app as a github repo?

**Comment by maxsnts** [comment]
Hello
Yes, i can.

I have created this repo:
https://github.com/maxsnts/YarpTest

The Readme.md has the shortest test case instructions

Thank you

**Comment by Tratcher** [comment]
Ok, I understand the issue now. Accessing `context.Request.Form` consumes the request body, and that body is not buffered by default so YARP can no longer proxy the request correctly. It gets hung on the destination app because the request includes a Content-Length header but no longer has a body. The request will eventually time out and abort.

Adding buffering allows the request to be proxied:
```
            if (context.Request.HasFormContentType)
            {
                context.Request.EnableBuffering();
                if (context.Request.Form.Count > 0)
                {
                    //...
                }
                context.Request.Body.Position = 0;
            }
```
or in Startup.ConfigureServices:
```
services.Configure<FormOptions>(o => o.BufferBody = true);
```

YARP or HttpClient could better report this situation by validating that the bytes written match the content-length, else abort the request.

**Comment by samsp-msft** [comment]
Can we detect if we are not at the beginning of the response body, or can't skip to it and then throw an exception? Yeah I know that can be unperformant, but its a good way to report back out to logging systems etc that there is a problem, and should show up for developers.

**Comment by Tratcher** [comment]
> Can we detect if we are not at the beginning of the response body, or can't skip to it and then throw an exception? Yeah I know that can be unperformant, but its a good way to report back out to logging systems etc that there is a problem, and should show up for developers.

Unfortunately without buffering there's no flag that lets you know the request body has been consumed.

**Comment by maxsnts** [comment]
Hi

Buffering may not be the preferred/transparent solution but it solves my problem, so thank you very much.

I guess that an Exception with a explicit message would be better that a hang.
At least would point people to the way to solve it.

Do you want to keep the issue open or is this a closed topic?
Can i delete de test Repo?

Thank you

**Comment by Tratcher** [comment]
Let's keep this issue open so we can decide if we need to add a mitigation.

You can delete the test repo. Thanks.

**Comment by karelz** [comment]
Triage:
- HttpClient should count bytes in request and abort / throw if it does not match Content-Length provided upfront - .NET 7.0 - @MihaZupan can you please file a big?
- We should fix it in YARP as well for 6.0 and earlier
